crates/telio-starcast/src/starcast_peer.rs

```diff
@@ -0,0 +1,312 @@
+use x25519_dalek::{PublicKey as PublicKeyDalek, StaticSecret};
+
+const MAX_PACKET: usize = 2048;
+
+#[derive(Debug, thiserror::Error)]
+/// Custom `UdpProxy` error
+pub enum Error {
+    /// `StarcastPeer` Not Configured Error
+    #[error("Starcast peer is not configured")]
+    NotConfigured,
+    /// `StarcastPeer` `IoError`
+    #[error("Starcast peer IoError")]
+    IoError(#[from] std::io::Error),
+    /// Task execution failed
+    #[error(transparent)]
+    Task(#[from] telio_task::ExecError),
+    /// IP network failed
+    #[error("Failed to parse IP network")]
+    IpNetworkError(#[from] ipnetwork::IpNetworkError),
+    /// Send failed
+    #[error(transparent)]
+    Send(#[from] SendTimeoutError<(PublicKey, DataMsg)>),
+    /// Tunnel creation failed
+    #[error("Tunnel creation failed {0}")]
+    TunnNewFail(&'static str),
+}
+
+// TODO this should probably be moved to a higher level file
+
+/// The actual starcast peer containing the task which runs the wait loop for this whole component.
+pub struct StarcastPeer {
+    task: Task<State>,
+}
+
+/// Starcast peer configuration.
+pub struct Config {
+    /// Public key of the starcast virtual peer.
+    public_key: PublicKey,
+    /// Port number through which the virtual starcast peer shall communicate with WireGuar.
+    wg_port: u16,
+}
+
+/// State for the starcast peer runtime where all the "working" members are stored and operated on.
+struct State {
+    /// Allowed IPs of the starcast peer.
+    allowed_ips: Vec<IpNetwork>,
+    /// Interval at which the tunnel timers will be updated.
+    wg_timer_tick_interval: Interval,
+    /// Channel for sending decapsulated packets to the multicaster and receiving packets for encapsulation from it.
+    channel: Chan<Vec<u8>>,
+    /// Local socket for sending and receiving packets from WireGuard.
+    socket: Option<UdpSocket>,
+    /// WireGuard socket address.
+    wg_sock_addr: Option<SocketAddr>,
+    /// WireGuard encryption tunnel for encapsulation/decapsulation of packets.
+    tunnel: Option<Tunn>,
+    /// Flag indicating that there are packets in the encryption tunnel internal buffer that need to be processed
+    /// before new packets can be received from the socket and decapsulated.
+    packets_present_in_tunnel: bool,
+    // TODO need to recheck if there are no issues with these buffers:
+    /// Buffer for outbound packets.
+    sending_buffer: Vec<u8>,
+    /// Buffer for inbound packets.
+    receiving_buffer: Vec<u8>,
+}
+
+impl State {
+    async fn configure(&mut self, config: Config) -> Result<u16, Error> {
+        let socket = UdpSocket::bind(SocketAddr::from(([127, 0, 0, 1], config.wg_port))).await?;
+        let actual_port = socket.local_addr()?.port();
+        self.socket = Some(socket);
+
+        self.wg_sock_addr = Some(SocketAddr::from(([127, 0, 0, 1], config.wg_port)));
+
+        self.tunnel = Some(
+            Tunn::new(
+                StaticSecret::from(SecretKey::gen().into_bytes()),
```

Comment:
It would be a smart to keep this key stable arround configs

Reply:
Fixed